x/valset-pref/validator_set.go

```diff
@@ -29,6 +29,92 @@ func (k Keeper) SetupValidatorSetPreference(ctx sdk.Context, delegator string, p
 	return nil
 }
 
+// DelegateToValidatorSet delegates to a delegators existing validator-set.
+// For ex: delegate 10osmo with validator-set {ValA -> 0.5, ValB -> 0.3, ValC -> 0.2}
+// our delegate logic would attempt to delegate 5osmo to A , 2osmo to B, 3osmo to C
+func (k Keeper) DelegateToValidatorSet(ctx sdk.Context, delegatorAddr string, coin sdk.Coin) error {
+	// get the existing validator set preference from store
+	existingSet, found := k.GetValidatorSetPreference(ctx, delegatorAddr)
+	if !found {
+		return fmt.Errorf("user %s doesn't have validator set", delegatorAddr)
+	}
+
+	delegator, err := sdk.AccAddressFromBech32(delegatorAddr)
+	if err != nil {
+		return err
+	}
+
+	// loop through the validatorSetPreference and delegate the proportion of the tokens based on weights
+	for _, val := range existingSet.Preferences {
+		_, validator, err := k.getValAddrAndVal(ctx, val.ValOperAddress)
+		if err != nil {
+			return err
+		}
+
+		// tokenAmt takes the amount to delegate, calculated by {val_distribution_weight * tokenAmt}
+		tokenAmt := val.Weight.Mul(coin.Amount.ToDec()).TruncateInt()
+
+		// TODO: What happens here if validator is jailed, tombstoned, or unbonding
```

Comment:
I think the only one we still are curious about is tombstoned

Reply:
TODO: change comments